src/main/kotlin/org/rust/ide/intentions/ExtractEnumVariantIntention.kt

```diff
@@ -0,0 +1,210 @@
+        val whereClause = enum.whereClause
+        val name = ctx.variant.name ?: return
+
+        val struct = element.createStruct(name, typeParameters, whereClause?.text ?: "")
+        val inserted = enum.parent.addBefore(struct, enum) as RsStructItem
+        val visibility = element.visibility
+
+        for (usage in ReferencesSearch
+            .search(ctx.variant)) {
+            element.replaceUsage(usage.element, name)
+        }
+
+        val newVariant = factory.createTupleFields(listOf(RsPsiFactory.TupleField(visibility, inserted.declaredType)))
+        val replaced = element.toBeReplaced.replace(newVariant)
+
+        offerStructRename(project, editor, inserted, replaced)
+    }
+
+    private fun filterTypeParameters(references: List<RsTypeReference>, parameters: RsTypeParameterList?): String {
+        return if (parameters != null) {
+            val lifetimes = parameters.lifetimeParameterList.filter { lifetime -> references.any { matchesLifetime(it, lifetime) } }
+            val typeParameters = parameters.typeParameterList.filter { param -> references.any { matchesParameter(it, param) } }
+            val filtered = (lifetimes + typeParameters)
+            if (filtered.isNotEmpty()) {
+                (lifetimes + typeParameters).joinToString(", ", prefix = "<", postfix = ">") { it.text }
+            } else {
+                ""
+            }
+        } else {
+            ""
+        }
+    }
+
+    private fun matchesParameter(ref: RsTypeReference, parameter: RsTypeParameter): Boolean {
+        val type = ref.type
+        return type.visitWith(HasTypeParameterVisitor(parameter))
+    }
+
+    private fun matchesLifetime(ref: RsTypeReference, parameter: RsLifetimeParameter): Boolean {
+        val type = ref.type
+        return type.visitWith(HasLifetimeParameterVisitor(parameter))
+    }
+
+    private fun offerStructRename(project: Project,
+                                  editor: Editor,
+                                  inserted: RsStructItem,
+                                  replaced: PsiElement) {
+        val range = inserted.identifier?.textRange ?: return
+        editor.caretModel.moveToOffset(range.startOffset)
+
+        PsiDocumentManager.getInstance(project).doPostponedOperationsAndUnblockDocument(editor.document)
+        RsInPlaceVariableIntroducer(inserted, editor, project, "Choose struct name", arrayOf(replaced))
+            .performInplaceRefactoring(LinkedHashSet())
+    }
+
+    private fun createElement(variant: RsEnumVariant, factory: RsPsiFactory): VariantElement {
+        val tupleFields = variant.tupleFields
+        val blockFields = variant.blockFields
+
+        return when {
+            tupleFields != null -> TupleVariant(tupleFields, factory)
+            blockFields != null -> StructVariant(blockFields, factory)
+            else -> error("unreachable")
+        }
+    }
+}
+
+private sealed class VariantElement(val toBeReplaced: PsiElement, val factory: RsPsiFactory) {
```

Comment:
Minor
Maybe it's better to create `factory` from `toBeReplaced`?
```kotlin
protected val factory: RsPsiFactory get() = RsPsiFactory(toBeReplaced.project)
```

Reply:
Right. For some reason I thought that the factory is expensive to create.